### ISSUE TITLE
No systray icon if autostarted by KDE Plasma

### ISSUE THREAD

**Issue by bam80** [telegramdesktop/tdesktop#16833]
Cross-linked from downstream issue:
https://bugs.kde.org/show_bug.cgi?id=425315

The bug was closed in KDE as UPSTREAM, please reopen if we are wrong:
https://bugs.kde.org/show_bug.cgi?id=425315#c28
> From the logs:
> 
> :1.230 is telegram
> It makes a DBus call to kded asking if we have any SNI hosts (i.e the system tray)
> KDED is up and ready and available and the call to the SNI watcher works.
> But there are no hosts as plasma isn't up yet so it returns false
> We later emit that a host is registered, but at this point telegram doesn't care.
> 
> From a plasma side that looks like everything is working correctly. Other than moving app restore until after plasma is fully up, there's not much we can do.

**Comment by ilya-fedin** [comment]
@Aokromes can you close the issue since template was removed?

**Comment by ilya-fedin** [comment]
Yeah, it's needed, kde tracker doesn't have all the information template requests. And I think you still filed bug report in wrong place, tray implementation is inside of Qt and it doesn't support appearing tray host on the fly. So, either KDE should 'moving app restore until after plasma is fully up' or Qt should rewrite it's tray implementation.

https://github.com/qt/qtbase/blob/5.15/src/platformsupport/themes/genericunix/qgenericunixthemes.cpp#L104-L115

**Comment by bam80** [comment]
> tray implementation is inside of Qt

I thought you have custom SNI implementation, or am I messing the things?
https://github.com/telegramdesktop/tdesktop/issues/8708#issuecomment-702788836

**Comment by ilya-fedin** [comment]
Custom SNI implementation subscribes to StatusNotifierHostRegistered: https://github.com/telegramdesktop/tdesktop/blob/dev/Telegram/SourceFiles/platform/linux/main_window_linux.cpp#L826-L845

**Comment by bam80** [comment]
I see, could this line from the log be a reason the subscription doesn't work?
```
dbus-daemon[2055]: apparmor="DENIED" operation="dbus_signal"  bus="session" path="/StatusNotifierWatcher" interface="org.kde.StatusNotifierWatcher" member="StatusNotifierHostRegistered" name=":1.28" mask="receive" pid=2353 label="snap.telegram-desktop.telegram-desktop" peer_pid=2225 peer_label="unconfined"
```
https://bugsfiles.kde.org/attachment.cgi?id=140801

**Comment by ilya-fedin** [comment]
Yeah

**Comment by bam80** [comment]
Thanks. Supposedly it's snap packaging issue then.
I'll check native binary installation and Flatpak.
If you have the solution for Snap, please also share.

**Comment by ilya-fedin** [comment]
> If you have the solution for Snap, please also share.

If they block that signal then there's no solution other than pleasure KDE developers to start applications after all services are started, like all other DEs do

**Comment by bam80** [comment]
I'm not aware but maybe there is a way to fine-tune the sandbox for Snap to allow some signals.
If not it become a more serious issue and maybe need to be reconsidered on Plasma side indeed.

**Comment by ilya-fedin** [comment]
> I'm not aware but maybe there is a way to fine-tune the sandbox for Snap to allow some signals.

No, unlike in flatpak, you can't get random permissions in snap. They have predefined sets of apparmor rules and they connect some 'slots' only after review by Canonical employees. Tray support is in desktop-legacy slot.
https://github.com/snapcore/snapd/blob/master/interfaces/builtin/desktop_legacy.go

I don't know where the right place to ask for a new rule, I tried once to ask to add permission to launch activatable notification daemons on snapcraft forum (forum.snapcraft.io), but no luck.

There's also a bugtracker on launchpad (https://bugs.launchpad.net/ubuntu/+source/snapd/+bugs), but I don't have launchpad account.

**Comment by bam80** [comment]
Thanks, at least we know it's possible for Snap in principle now.
I opened the desktop-legacy slot request here:
https://forum.snapcraft.io/t/desktop-legacy-slot-request-for-the-telegram-desktop-snap-tray-support/26091

**Comment by ilya-fedin** [comment]
tdesktop already has desktop-legacy host. The problem is it doesn't allow StatusNotifierHostRegistered signal. I.e. we need slot to be changed in snapd source code, not to be connected.

**Comment by bam80** [comment]
Got it. Could you create a topic on their forum so we could track it there?

**Comment by ilya-fedin** [comment]
I would honestly say I don't want to do that after their (lack of) reaction to my latest topic

**Comment by bam80** [comment]
That's sad, maybe we would have better luck if we just provide the PR? Especially if the change is not controversial.

**Comment by ilya-fedin** [comment]
> Especially if the change is not controversial.

I guess adding new permission for thousand of apps is always controversial. I definitely won't PR such a thing.

**Comment by bam80** [comment]
I wish we don't stick here, but so far I couldn't convince Plasma devs to do any changes in it's code.
Having that, to be honest the problem seems to be on Snap side and not anywhere else.

**Comment by ilya-fedin** [comment]
Qt's and Electron's SNI implementations don't seem to work on the fly, so they still need to move application start after all services started in order to work correctly with them.

**Comment by bam80** [comment]
For some reason I don't have this problem with Discord which is Electron app AFAIK.
Either way, if StatusNotifierHostRegistered signal is not Plasma-specific we probably can say it's Toolkit's problem?

**Comment by ilya-fedin** [comment]
Entire spec is Plasma-specific (invented by Plasma)

**Comment by ilya-fedin** [comment]
I can say no other DE does that weird thing with deferred tray start. Only Plasma does that for some reason, with both tray & notifications and that caused a lot of pain since it was non-trivial to rewrite tdesktop's notification subsystem to be able to switch notification backend on the fly.

**Comment by bam80** [comment]
> Entire spec is Plasma-specific (invented by Plasma)

Still it's a spec already. But looks like we are deviating from the topic.

**Comment by ilya-fedin** [comment]
> Still it's a spec already.

Formally it was never standardized, it's a de-facto standard.

**Comment by bam80** [comment]
Can Flatpak suffer, too?
https://github.com/flathub/org.goldendict.GoldenDict/issues/4
